falcon_test.go

```diff
@@ -109,28 +108,232 @@ func TestFalcon(t *testing.T) {
+	}
+
+	ctSignature, err := sig.ConvertToCT()
+	if err != nil {
+		t.Fatalf("failed to verify message. err message: %s", err)
+	}
+
+	err = pub.VerifyCTSignature(ctSignature, nil)
+	if err != nil {
+		t.Fatalf("failed to verify message. err message: %s", err)
+	}
+
+	err = pub.VerifyCTSignature(ctSignature, []byte{})
+	if err != nil {
+		t.Fatalf("failed to verify message. err message: %s", err)
+	}
+}
+
+func TestFalconGenerateKeysDifferentSeed(t *testing.T) {
+	seed := make([]byte, 64)
+	rand.Read(seed)
+
+	pub, sk, err := GenerateKey(seed)
+	if err != nil {
+		t.Fatalf("failed to generate keys. err message: %s", err)
+	}
+
+	rand.Read(seed)
+	pub2, sk2, err := GenerateKey(seed)
+	if err != nil {
+		t.Fatalf("failed to generate keys. err message: %s", err)
+	}
+
+	if pub == pub2 {
+		t.Fatalf("public keys are the same")
+	}
+
+	if sk == sk2 {
+		t.Fatalf("private keys are the same")
+	}
+}
+
+func TestFalconNilSignature(t *testing.T) {
+	seed := make([]byte, 64)
+	rand.Read(seed)
+
+	pub, _, err := GenerateKey(seed)
+	if err != nil {
+		t.Fatalf("failed to generate keys. err message: %s", err)
+	}
+
+	msg := make([]byte, 500)
+	rand.Read(msg)
+
+	err = pub.Verify(nil, msg)
+	if err == nil {
+		t.Fatalf("verification succeeded. should have failed.")
+	}
+
+	err = pub.Verify([]byte{}, msg)
+	if err == nil {
+		t.Fatalf("verification succeeded. should have failed.")
+	}
+}
+
+func TestFalconNilSeed(t *testing.T) {
+	_, _, err := GenerateKey(nil)
+	if err != nil {
+		t.Fatalf("failed to generate keys. err message: %s", err)
+	}
+
+	_, _, err = GenerateKey([]byte{})
+	if err != nil {
+		t.Fatalf("failed to generate keys. err message: %s", err)
+	}
+}
+
+func TestSaltedVersions(t *testing.T) {
+	emptyCTSig := CTSignature{}
+	if emptyCTSig.SaltVersion() != -1 {
+		t.Fatalf("exptected salt value to be error")
+	}
+
+	emptyCompressSig := CompressedSignature{}
+	if emptyCompressSig.SaltVersion() != -1 {
+		t.Fatalf("exptected salt value to be error")
+	}
+
+	emptyCompressSig = []byte{0x0}
+	if emptyCompressSig.SaltVersion() != -1 {
+		t.Fatalf("exptected salt value to be error")
+	}
+}
+
+func BenchmarkFalconKeyGen(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		var seed [48]byte
+		rand.Read(seed[:])
+		GenerateKey(seed[:])
+	}
+}
+
+func BenchmarkFalconSignCompressed(b *testing.B) {
+	_, sk, err := GenerateKey([]byte("seed"))
+	if err != nil {
+		b.Fatalf("GenerateKey failed")
+	}
+
+	strs := make([][64]byte, b.N)
+	for i := 0; i < b.N; i++ {
+		var msg [64]byte
+		rand.Read(msg[:])
+		strs[i] = msg
+	}
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		sk.SignCompressed(strs[i][:])
+	}
+}
+
+func BenchmarkFalconVerify(b *testing.B) {
+	pk, sk, err := GenerateKey([]byte("seed"))
+	if err != nil {
+		b.Error(err, "GenerateKey failed")
+	}
+
+	strs := make([][64]byte, b.N)
+	sigs := make([]CompressedSignature, b.N)
+	for i := 0; i < b.N; i++ {
+		var msg [64]byte
+		rand.Read(msg[:])
+		strs[i] = msg
+		sigs[i], _ = sk.SignCompressed(msg[:])
```

Comment:
since it's in the init part of the benchmark, I would test the returned error code and fail as needed.